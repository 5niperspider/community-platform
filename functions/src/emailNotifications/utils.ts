import { NotificationType } from '../../../src/models'
import { CONFIG } from '../config/config'
import { INotification } from '../models'
import {
  PP_PROJECT_IMAGE,
  PK_PROJECT_IMAGE,
  PP_PROJECT_NAME,
  PK_PROJECT_NAME,
} from './consts'

export const SITE_URL = CONFIG.deployment.site_url

export const getProjectImageSrc = () => {
  switch (SITE_URL) {
    case 'https://dev.onearmy.world':
    case 'https://community.preciousplastic.com':
      return PP_PROJECT_IMAGE
    case 'https://dev.community.projectkamp.com':
    case 'https://community.projectkamp.com':
      return PK_PROJECT_IMAGE
  }
}

export const getProjectName = () => {
  switch (SITE_URL) {
    case 'https://dev.onearmy.world':
    case 'https://community.preciousplastic.com':
      return PP_PROJECT_NAME
    case 'https://dev.community.projectkamp.com':
    case 'https://community.projectkamp.com':
      return PK_PROJECT_NAME
  }
}

export const getUserLink = (displayName: string, userName: string) =>
  `<a href='${SITE_URL}/u/${userName}'>${displayName}</a>`

export const getResourceLink = (
  notificationType: NotificationType,
  relevantUrl: string,
) => {
  let resourceLabel: string
  switch (notificationType) {
    case 'new_comment':
    case 'howto_useful':
    case 'howto_mention':
      resourceLabel = 'how-to'
    case 'new_comment_research':
    case 'research_useful':
    case 'research_mention':
    case 'research_update':
      resourceLabel = 'research'
  }
  return `<a href='${SITE_URL}${relevantUrl}'>${resourceLabel}</a>`
}

export const getCommentListItem = (notification: INotification) => `
<p>
    New comment on your ${getResourceLink(
      notification.type,
      notification.relevantUrl,
    )} by ${getUserLink(
  notification.triggeredBy.displayName,
  notification.triggeredBy.userName,
)}
</p>`

export const getMentionListItem = (notification: INotification) => `
<p>
  ${getUserLink(
    notification.triggeredBy.displayName,
    notification.triggeredBy.userName,
  )} mentioned you in this ${getResourceLink(
  notification.type,
  notification.relevantUrl,
)}
</p>`

export const getUsefulListItem = (notification: INotification) => `
<p>
    ${getUserLink(
      notification.triggeredBy.displayName,
      notification.triggeredBy.userName,
    )} found your ${getResourceLink(
  notification.type,
  notification.relevantUrl,
)}
useful
</p>
`

export const isCommentNotification = (notification: INotification) =>
  ['new_comment_research', 'new_comment'].includes(notification.type)

export const isMentionNotification = (notification: INotification) =>
  ['research_mention', 'howto_mention'].includes(notification.type)

export const isUsefulNotification = (notification: INotification) =>
  ['research_useful', 'howto_useful'].includes(notification.type)
