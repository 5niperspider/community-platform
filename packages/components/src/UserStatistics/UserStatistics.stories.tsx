import type { StoryFn, Meta } from '@storybook/react'
import { UserStatistics } from './UserStatistics'
import { faker } from '@faker-js/faker'

export default {
  title: 'Components/UserStatistics',
  component: UserStatistics,
} as Meta<typeof UserStatistics>

export const Default: StoryFn<typeof UserStatistics> = () => (
  <UserStatistics
    userName={faker.internet.userName()}
    country={faker.address.country()}
    isVerified={true}
    isSupporter={true}
    howtoCount={faker.datatype.number()}
    eventCount={faker.datatype.number()}
  />
)
