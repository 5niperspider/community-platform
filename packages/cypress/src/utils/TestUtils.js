"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setIsPreciousPlastic = exports.generateNewUserDetails = exports.DbCollectionName = exports.generateAlphaNumeric = exports.Page = void 0;
var Page;
(function (Page) {
    Page["HOWTO"] = "/how-to";
    Page["ACADEMY"] = "/academy";
    Page["SETTINGS"] = "/settings";
})(Page || (exports.Page = Page = {}));
var generateAlphaNumeric = function (length) {
    var result = '';
    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for (var i = 0; i < length; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
};
exports.generateAlphaNumeric = generateAlphaNumeric;
var DbCollectionName;
(function (DbCollectionName) {
    DbCollectionName["users"] = "users";
    DbCollectionName["howtos"] = "howtos";
})(DbCollectionName || (exports.DbCollectionName = DbCollectionName = {}));
var generateNewUserDetails = function () {
    return {
        username: "CI_".concat((0, exports.generateAlphaNumeric)(5)).toLocaleLowerCase(),
        email: "CI_".concat((0, exports.generateAlphaNumeric)(5), "@test.com").toLocaleLowerCase(),
        password: 'test1234',
    };
};
exports.generateNewUserDetails = generateNewUserDetails;
var setIsPreciousPlastic = function () {
    return localStorage.setItem('platformTheme', 'precious-plastic');
};
exports.setIsPreciousPlastic = setIsPreciousPlastic;
