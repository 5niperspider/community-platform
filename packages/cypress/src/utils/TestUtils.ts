export enum Page {
  HOWTO = '/how-to',
  ACADEMY = '/academy',
  SETTINGS = '/settings',
}

export const generateAlphaNumeric = (length: number) => {
  let result = ''
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength))
  }
  return result
}

export enum DbCollectionName {
  users = 'users',
  howtos = 'howtos',
}

export const fillSignupForm = (user) => {
  const { username, email, password } = user

  cy.get('[data-cy=username]').clear().type(username)
  cy.get('[data-cy=email]').clear().type(email)
  cy.get('[data-cy=password]').clear().type(password)
  cy.get('[data-cy=confirm-password]').clear().type(password)
  cy.get('[data-cy=consent]').check()
}

export const signUpNewUser = (newUser?) => {
  const user = newUser || generateNewUserDetails()

  cy.step(`Sign up new user - ${user.username}`)
  cy.visit('/sign-up')
  fillSignupForm(user)
  cy.get('[data-cy=submit]').click()
  cy.url().should('include', 'sign-up-message')
}

export const generateNewUserDetails = () => {
  return {
    username: `CI_${generateAlphaNumeric(5)}`.toLocaleLowerCase(),
    email: `CI_${generateAlphaNumeric(5)}@test.com`.toLocaleLowerCase(),
    password: 'test1234',
  }
}
