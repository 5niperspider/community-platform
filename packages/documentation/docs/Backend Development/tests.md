---
id: tests
title: Tests
---

Backend tests are written with [Jest](https://jestjs.io/docs/getting-started) and executed against a set of emulators for core services

## Prerequisites

You will need to support the running of [Firebase Emulators](https://firebase.google.com/docs/emulator-suite/install_and_configure#install_the_local_emulator_suite)

## Running Locally

```
yarn workspace functions test:watch
```

This will startup the firebase emulator suite, compile functions, and watch for any changes to function or spec test files

:::info  
The emulator console can be viewd on [localhost:4001](http://localhost:4001). All data will be cleared on script termination
:::

To run just a single test the interactive prompts can be used to provide a matching filename path, e.g.

```sh
$ yarn workspace functions test:watch

$ Watch Usage
> Press p to filter by a filename regex pattern

$ pattern â€º mySpec

```

This will automatically watch for changes to any files that match the regular expression `/mySpec/`

## Writing Tests

Tests are written in [Jest](https://jestjs.io/docs/getting-started), with filenames ending `.spec.ts`

## Utilities and Examples

### Mocking

TODO - share examples of mocking methods to avoid need for full function execution

### Executing Functions Directly

It is possible to directly execute any function within the test environment without its required trigger.
A utility `FirebaseEmulatedTest` class is used to wrap the function invocation so that it can be used async, e.g.

_myFunction.ts_

```ts
import functions from 'firebase-functions'

exports.default = functions.firestore
  .document(`mockEndpoint/{id}`)
  .onUpdate((change) => {
    return doSomething(change)
  })
```

_myFunction.spec.ts_

```ts
import { FirebaseEmulatedTest } from '../test/Firebase/emulator'
import myFunction from './myFunction.ts'

const beforeData = { field: 'initialValue' }
const afterData = { field: 'changedValue' }

const change = FirebaseEmulatedTest.mockFirestoreChangeObject(
  beforeData,
  afterData,
  'mockEndpoint',
  'mockDocId',
)

await FirebaseEmulatedTest.run(myFunction, change)
```

Additional utilities can also be used to provide a mock of the triggering context. In the example above a mock `before` and `after` data snapshot is created to mimic the firestore document change trigger used by the function

### Seed and Teardown Data

The test utilities also have methods for seeding and clearing the Firestore DB if required for tests.

_myFunction.spec.ts_

```ts
beforeEach(async () => {
  await FirebaseEmulatedTest.seedFirestoreDB('users', [{ _id: 'user1 ' }])
})
afterEach(async () => {
  await FirebaseEmulatedTest.clearFirestoreDB()
})
```

:::caution  
If your test function interacts directly with an empty database endpoint it may throw an error  
(various issues on github, should review in the future)
:::

The easiest workaround is to seed the endpoint without any docs, which will allow the emulators to return an empty array from queries.

```
await FirebaseEmulatedTest.seedFirestoreDB('empty_endpoint',[])
```
