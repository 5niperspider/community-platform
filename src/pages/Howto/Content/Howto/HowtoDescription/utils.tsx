interface localStorageExpiry {
  howtoID: string
  expiry: number
}

export const retrieveLocalStorageArray = ():
  | localStorageExpiry[]
  | undefined => {
  const downloadCooldownArray: string | null =
    localStorage.getItem('downloadCooldown')
  if (typeof downloadCooldownArray === 'string') {
    return JSON.parse(downloadCooldownArray)
  }
}

export const retrieveHowtoDownloadCooldown = (
  howtoID: string,
): localStorageExpiry | undefined => {
  const downloadCooldownArray = retrieveLocalStorageArray()
  if (downloadCooldownArray) {
    return downloadCooldownArray.find((elem) => elem.howtoID === howtoID)
  }
}

export const isHowtoDownloadCooldownExpired = (
  howtoCooldown: localStorageExpiry,
): boolean => {
  const now = new Date()
  if (now.getTime() > howtoCooldown.expiry) {
    return true
  } else {
    return false
  }
}
export const createHowtoExpiryObject = (
  howtoID: string,
): localStorageExpiry => {
  const now = new Date()
  const twelveHoursInMiliseconds = 12 * 60 * 60 * 1000
  return {
    howtoID: howtoID,
    expiry: now.getTime() + twelveHoursInMiliseconds,
  }
}

export const addHowtoDownloadCooldown = (howtoID: string) => {
  let downloadCooldownArray = retrieveLocalStorageArray()
  const expiryObject = createHowtoExpiryObject(howtoID)

  if (!downloadCooldownArray) {
    // create array for localstorage if none exists
    downloadCooldownArray = []
    downloadCooldownArray.push(expiryObject)
  } else {
    downloadCooldownArray.push(expiryObject)
  }
  localStorage.setItem(
    'downloadCooldown',
    JSON.stringify(downloadCooldownArray),
  )
}

export const updateHowtoDownloadCooldown = (howtoID: string) => {
  const downloadCooldownArray = retrieveLocalStorageArray()
  const expiryObject = createHowtoExpiryObject(howtoID)
  const foundIndex = downloadCooldownArray!.findIndex(
    (elem) => elem.howtoID === howtoID,
  )

  downloadCooldownArray![foundIndex] = expiryObject
  localStorage.setItem(
    'downloadCooldown',
    JSON.stringify(downloadCooldownArray),
  )
}
