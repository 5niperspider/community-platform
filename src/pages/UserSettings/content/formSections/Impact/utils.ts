import { impactQuestions } from '../Impact/impactQuestions'

import type { IImpactDataField, IImpactYearFieldList } from 'src/models'
import type { IImpactQuestion } from '../Impact/impactQuestions'

export interface ImpactDataFieldInputs {
  [key: string]: IImpactQuestion
}

interface IInputs {
  [key: string]: IImpactDataField
}

export const transformImpactData = (
  fields: IImpactYearFieldList,
): ImpactDataFieldInputs => {
  const transformed = {} as ImpactDataFieldInputs

  Object.values(fields).forEach((field) => {
    const questionField = impactQuestions.find(
      (question) => question.label === field.label,
    )
    return (transformed[field.label] = {
      ...field,
      description: questionField?.description,
    } as IImpactQuestion)
  })

  return transformed
}

export const transformImpactInputs = (
  inputs: IInputs,
): IImpactYearFieldList => {
  const fields = [] as IImpactYearFieldList

  Object.keys(inputs).forEach((key) => {
    const field = inputs[key]
    const question = impactQuestions.find(({ label }) => label === key)
    if (field && question && field.value) {
      fields.push({
        label: key,
        value: field.value,
        isVisible:
          typeof field.isVisible === 'boolean' ? field.isVisible : true,
        ...(question.prefix && { prefix: question.prefix }),
        ...(question.suffix && { suffix: question.suffix }),
      })
    }
  })

  return fields
}
