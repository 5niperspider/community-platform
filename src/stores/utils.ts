import type { HowtoStore } from './Howto/howto.store'
import type { ResearchStore } from './Research/research.store'

export const storeHasAuthorFunctionality = (
  store: any,
): store is HowtoStore | ResearchStore => {
  return (
    store &&
    typeof store.selectedAuthor !== 'undefined' &&
    typeof store.updateSelectedAuthor === 'function'
  )
}
