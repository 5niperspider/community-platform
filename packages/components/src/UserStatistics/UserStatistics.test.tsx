import { render } from '../tests/utils'
import { Default } from './UserStatistics.stories'
import type { UserStatisticsProps } from './UserStatistics'

describe('UserStatistics', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Default {...(Default.args as UserStatisticsProps)} />,
    )

    expect(container).toMatchSnapshot()
  })
})
