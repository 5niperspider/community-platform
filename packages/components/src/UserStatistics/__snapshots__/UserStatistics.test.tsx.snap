// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserStatistics > renders correctly 1`] = `
<div>
  <div
    class="css-1m50i6j-Box"
  >
    <div
      class="css-n3dlyc-Box"
    >
      <div
        class="css-i1ihic-Box"
      >
        <div
          class="css-1f8r5dj"
          size="22"
        >
          <img
            alt="icon"
            src="/assets/icons/icon-verified-badge.svg"
            style="max-width: 100%;"
          />
        </div>
        <div
          class="css-tj1vn5-Box"
        >
          Verified
        </div>
      </div>
      <a
        class="css-17ig1b0-Box"
        href="/map/#Test User"
      >
        <div
          class="css-i1ihic-Box"
        >
          <div
            class="css-1f8r5dj"
            size="22"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="width: 22px; height: 22px;"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
              />
            </svg>
          </div>
          <div
            class="css-tj1vn5-Box"
          >
            Greenland
          </div>
        </div>
      </a>
      <div
        class="css-i1ihic-Box"
      >
        <div
          class="css-1f8r5dj"
          size="22"
        >
          <img
            alt="icon"
            src="/assets/icons/supporter.svg"
            style="max-width: 100%;"
          />
        </div>
        <div
          class="css-tj1vn5-Box"
        >
          <a
            class="css-17ig1b0-Box"
            href="https://www.patreon.com/one_army"
            rel="noopener noreferrer"
            target="_blank"
          >
            Supporter
          </a>
        </div>
      </div>
      <div
        class="css-i1ihic-Box"
      >
        <div
          class="css-1bhrp1g-Box"
        />
        Useful: 
        20
      </div>
      <div
        class="css-i1ihic-Box"
      >
        <div
          class="css-19048r3-Box"
        />
        How‑to: 
        10
      </div>
      <div
        class="css-i1ihic-Box"
      >
        <div
          class="css-8n4scg-Box"
        />
        Events: 
        4
      </div>
    </div>
  </div>
</div>
`;
